### PR TITLE
feat(authentication): manage openid role mapping

### PR DESCRIPTION
**Pull request by wtermellil**
## Description

**PLEASE MAKE SURE THAT THE BRANCH PR INCLUDES JIRA TICKET ID** (_for centreon-internal_)

New section on OpenID configuration named Role mapping

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.04.x
- [ ] 21.10.x
- [ ] 22.04.x
- [x] 22.10.x (master)

<h2> How this pull request can be tested ? </h2>

Configure your OpenID

#### Community contributors & Centreon team

- [ ] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: src/Core/Security/ProviderConfiguration/Application/OpenId/UseCase/FindOpenIdConfiguration/FindOpenIdConfigurationResponse.php [hunk 1]
```diff
@@ -135,6 +135,11 @@ class FindOpenIdConfigurationResponse
      */
     public array $authorizationRules = [];
 
+    /**
+     * @var array
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Application/OpenId/UseCase/FindOpenIdConfiguration/FindOpenIdConfigurationResponse.php]: describe type of data in array

File: src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/ACLConditions.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+     */
+    public function isEnabled(): bool
+    {
+        return $this->isEnabled;
+    }
+
+    /**
+     * @return bool
+     */
+    public function isApplyOnlyFirstRole(): bool
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/ACLConditions.php]:
```suggestion
    public function isApplyingOnlyFirstRole(): bool
```

or

```suggestion
    public function ApplyOnlyFirstRole(): bool
```

File: src/Core/Security/ProviderConfiguration/Infrastructure/OpenId/Api/UpdateOpenIdConfiguration/UpdateOpenIdConfigurationSchema.json [hunk 3]
```diff
@@ -1,157 +1,267 @@
...
+          "claim_value": {
+            "type": "string"
           },
-          "name": {
+          "access_group_id": {
+            "type": "integer"
+          }
+        }
+      }
+    },
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Infrastructure/OpenId/Api/UpdateOpenIdConfiguration/UpdateOpenIdConfigurationSchema.json]: I think this should be remove as it is now in roles_mapping

File: src/Core/Security/ProviderConfiguration/Domain/OpenId/Exceptions/InvalidEndpointException.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Core\Security\ProviderConfiguration\Domain\OpenId\Exceptions;
+
+class InvalidEndpointException extends OpenIdConfigurationException
+{
+    /**
+     * @return static
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Domain/OpenId/Exceptions/InvalidEndpointException.php]:
```suggestion
     * @return self
```

File: src/Core/Security/ProviderConfiguration/Domain/OpenId/Exceptions/InvalidEndpointException.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    public static function invalidType(): self
+    {
+        return new self(_('Invalid endpoint type specified'));
+    }
+
+    /**
+     * @return static
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Domain/OpenId/Exceptions/InvalidEndpointException.php]:
```suggestion
     * @return self
```

File: src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/ACLConditions.php [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+ * ACL condition block regarding OpenID.
+ */
+class ACLConditions
+{
+    /**
+     * @param bool $isEnabled
+     * @param bool $applyOnlyFirstRole
+     * @param string $attributePath
+     * @param EndpointCondition $endpoint
+     * @param array $relations
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/ACLConditions.php]: describe type of data in relations

File: src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/ACLConditions.php [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+     * @throws ACLConditionsException
+     */
+    public function __construct(
+        private bool $isEnabled,
+        private bool $applyOnlyFirstRole,
+        private string $attributePath,
+        private EndpointCondition $endpoint,
+        private array $relations = []
+    )
+    {
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/ACLConditions.php]:
```suggestion
    ) {
```

File: src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/CustomConfiguration.php [hunk 8]
```diff
@@ -289,6 +294,88 @@ public function getContactGroup(): ?ContactGroup
...
+        if (
+            filter_var($clientAddress, FILTER_VALIDATE_IP) === false
+            && filter_var($clientAddress, FILTER_VALIDATE_DOMAIN, FILTER_FLAG_HOSTNAME) === false
+        ) {
+            throw AssertionException::ipOrDomain(
+                $clientAddress,
+                'OpenIdCustomConfiguration::' . $fieldName
+            );
+        }
+    }
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/CustomConfiguration.php]:
```suggestion
```
those line should be remove as they are now in authentication conditions

**Review by jeremyjaouen — CHANGES_REQUESTED** [review]

File: src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/EndpointCondition.php [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+     */
+    private static array $allowedTypes = [
+        self::INTROSPECTION,
+        self::USER_INFORMATION,
+        self::CUSTOM
+    ];
+
+    /**
+     * @param string $type
+     * @param string $url
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/EndpointCondition.php]:
```suggestion
     * @param string|null $url
```

File: src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/EndpointCondition.php [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+     * @throws InvalidEndpointException
+     */
+    public function __construct(
+        private string $type = self::INTROSPECTION,
+        private string $url = '',
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/EndpointCondition.php]:
```suggestion
        private ?string $url = null,
```

**Review by jeremyjaouen — APPROVED** [review]

File: src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/Endpoint.php [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+class Endpoint
+{
+    public const INTROSPECTION = 'introspection_endpoint';
+    public const USER_INFORMATION = 'user_information_endpoint';
+    public const CUSTOM = 'custom_endpoint';
+
+    /**
+     * @var string[]
+     */
+    private static array $allowedTypes = [
```
Comment (kduret) [inline src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/Endpoint.php]: any reason to use a static instead of a constant ?

Comment (wtermellil) [reply src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/Endpoint.php]: it's a mistake, I fix it now

File: src/Core/Security/ProviderConfiguration/Application/OpenId/UseCase/UpdateOpenIdConfiguration/UpdateOpenIdConfiguration.php [hunk 12]
```diff
@@ -184,6 +190,25 @@ private function createAuthorizationRules(array $authorizationRulesFromRequest):
         return $authorizationRules;
     }
 
+    /**
+     * @param array<string,bool|string|string[]|array<array{claim_value: string, access_group_id: int}>> $roles_mapping
+     * @return ACLConditions
+     * @throws \Throwable
+     */
+    private function createAclConditions(array $roles_mapping): ACLConditions
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Application/OpenId/UseCase/UpdateOpenIdConfiguration/UpdateOpenIdConfiguration.php]:
```suggestion
    private function createAclConditions(array $rolesMapping): ACLConditions
```

File: src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/Endpoint.php [hunk 13]
```diff
@@ -33,7 +33,7 @@ class Endpoint
     /**
      * @var string[]
      */
-    private static array $allowedTypes = [
+    private array $allowedTypes = [
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/Endpoint.php]:
```suggestion
    public const ALLOWED_TYPES = [
```

**Review by jeremyjaouen — APPROVED** [review]

File: lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 14]
```diff
@@ -17048,4 +17048,13 @@ msgid "Request for authentication conditions custom endpoint has failed"
...
+msgstr "Les conditions autorisées ne figurent pas dans les conditions du provider"
+
+msgid "The request for roles mapping on custom endpoint has failed"
+msgstr "La demande d'association de rôles a échoué en interrogeant l'url personalisée"
+
+msgid "Invalid roles mapping fetched from provider"
+msgstr "Les roles récupérés depuis le serveur d'identité sont incorrects"
+
+msgid "Role mapping conditions do not match"
+msgstr "Aucune correspondance trouvé pour le coupe roles / groupe d'accès"
```
Comment (kduret) [inline lang/fr_FR.UTF-8/LC_MESSAGES/messages.po]:
```suggestion
msgstr "Aucune correspondance trouvée pour le couple roles / groupe d'accès"
```

File: src/Core/Security/Authentication/Domain/Provider/OpenIdProvider.php [hunk 15]
```diff
@@ -32,24 +32,53 @@
...
+    {
+        $i = 0;
+        foreach ($array as $k => $v) {
+            if ($k !== $i++) {
+                return false;
+            }
+        }
+        return true;
+    }
+}
```
Comment (kduret) [inline src/Core/Security/Authentication/Domain/Provider/OpenIdProvider.php]: you can remove it now :)

File: src/Core/Security/Authentication/Domain/Provider/OpenIdProvider.php [hunk 16]
```diff
@@ -576,7 +615,7 @@ private function sendRequestForUserInformationEndpoint(): array
         $headers = [
             'Authorization' => "Bearer " . trim($this->providerToken->getToken())
         ];
-        /** @var CustomConfiguration $customConfiguration */
+         /** @var CustomConfiguration $customConfiguration */
```
Comment (kduret) [inline src/Core/Security/Authentication/Domain/Provider/OpenIdProvider.php]:
```suggestion
        /** @var CustomConfiguration $customConfiguration */
```

**Review by kduret — APPROVED** [review]

**Pull request merged**